python/ray/rllib/parallel.py

```diff
@@ -0,0 +1,245 @@
+from collections import namedtuple
+import os
+
+from tensorflow.python.client import timeline
+import tensorflow as tf
+
+
+class LocalSyncParallelOptimizer(object):
+  """Optimizer that runs in parallel across multiple local devices.
+
+  LocalSyncParallelOptimizer automatically splits up and loads training data
+  onto specified local devices (e.g. GPUs) with `load_data()`. During a call to
+  `optimize()`, the devices compute gradients over slices of the data in
+  parallel. The gradients are then averaged and applied to the shared weights.
+
+  The data loaded is pinned in device memory until the next call to
+  `load_data`, so you can make multiple passes (possibly in randomized order)
+  over the same data once loaded.
+
+  This is similar to tf.train.SyncReplicasOptimizer, but works within a single
+  TensorFlow graph, i.e. implements in-graph replicated training:
+
+    https://www.tensorflow.org/versions/r0.12/how_tos/distributed/
```

Comment:
let's point to https://www.tensorflow.org/api_docs/python/tf/train/SyncReplicasOptimizer here, the above link is from 0.12 which is pretty old by now